File: baekjoon/exhaustiveSearch/5014.py
import sys
from collections import deque
import copy

input = sys.stdin.readline

F, S, G, U, D = map(int,input().split())

isVisited = [0] * (F+1)

def bfs(start):
    q = deque([(start, 0)])
    visited = {start}
    while q:
        curFloor, cnt = q.popleft()

        # print(isVisited)
        # print(curFloor)
        # print(cnt)
        if curFloor == G:
            return cnt

        cnt += 1
        if curFloor + U <= F and curFloor + U not in visited:
            q.append((curFloor + U, cnt))
            visited.add(curFloor + U)
        if curFloor - D >= 1 and curFloor - D not in visited:
            q.append((curFloor - D, cnt))
            visited.add(curFloor - D)

    return "use the stairs"
#
# if U == D == 0:
#     print("use the stairs")
# elif S == G:
#     print(0)
# elif S < G and D == 0:
#     if (G - S) % U == 0:
#         print((G - S) // U)
#     else:
#         print("use the stairs")
# elif S < G and U == 0:
#     print("use the stairs")
# elif S > G and U == 0:
#     if (S - G) % D == 0:
#         print((S - G) // D)
#     else:
#         print("use the stairs")
# elif S > G and D == 0:
#     print("use the stairs")
# elif U == D:
#     if S < G:
#         if (G - S) % U == 0:
#             print((G - S) // U)
#         else:
#             print("use the stairs")
#     elif S > G:
#         if (S - G) % D == 0:
#             print((S - G) // D)
#         else:
#             print("use the stairs")
# else:
print(bfs(S))

File: baekjoon_js/exhaustiveSearch/5014.js
("use strict");

const fs = require("fs");
const stdin = (process.platform === "linux"
  ? fs.readFileSync("/dev/stdin").toString()
  : `10 1 10 2 1
`
).split("\n");

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const [F, S, G, U, D] = input().split(" ").map(Number);

let isVisited = Array(F + 1).fill(false);
let upDownBtn = [U, -D];
let isPossible = false;

console.log(BFS(S));

function BFS(start) {
  q = [];
  q.push([start, 0]);

  while (q.length > 0) {
    let [nowStair, nowCnt] = q.shift();

    if (nowStair === G) return nowCnt;
    nowCnt++;
    for (let i = 0; i < 2; i++) {
      if (
        nowStair + upDownBtn[i] > 0 &&
        nowStair + upDownBtn[i] < F + 1 &&
        isVisited[nowStair + upDownBtn[i]] === false
      ) {
        q.push([nowStair + upDownBtn[i], nowCnt]);
        isVisited[nowStair + upDownBtn[i]] = true;
      }
      //   console.log(nowStair, q, nowStair + upDownBtn[i]);
    }
  }

  return "use the stairs";
}
